spec/integration/apply_spec.rb

```diff
@@ -4,266 +4,288 @@
-          'puppet_library' => {
-            'plugin' => 'puppet_agent'
-          }
-        }
-      }, {
-        # These fail the puppet_agent::version check if they're fake. Seems
-        # like more effort than it's worth to mock them
-        'uri' => conn_uri('ssh'),
-        'name' => 'badparams',
-        'plugin_hooks' => {
-          'puppet_library' => {
-            'plugin' => 'task',
-            'task' => 'puppet_agent::install',
-            'parameters' => {
-              'collection' => 'The act or process of collecting.'
-            }
-          }
-        }
-      }, {
-        'uri' => conn_uri('ssh'),
-        'name' => 'badplugin',
-        'plugin_hooks' => {
-          'puppet_library' => {
-            'plugin' => 'what plugin?'
-          }
-        }
-      }] }
-    end
+    it 'succeeds with an empty hiera config' do
+      results = run_cli_json(%W[plan run prep -t #{targets}], project: project)
 
-    def task_plugin_inventory
-      { 'targets' => [{
-        'uri' => conn_uri('ssh'),
-        'plugin_hooks' => {
-          'puppet_library' => {
-            'plugin' => 'task',
-            'task' => 'puppet_agent::install',
-            'parameters' => { 'version' => '6.2.0' }
-          }
-        }
-      }],
-        'config' => root_config }
+      results.each do |result|
+        expect(result['status']).to eq('success')
+        expect(result.dig('value', 'report', 'resource_statuses')).to include(/Notify\[Hello .*\]/)
+      end
     end
+  end
 
-    after(:all) do
-      ssh_node = conn_uri('ssh', include_password: true)
-      uninstall([ssh_node, 'agent_targets'], inventory: agent_version_inventory)
-    end
+  describe 'over ssh', ssh: true do
+    let(:uri)      { conn_uri('ssh') }
+    let(:user)     { conn_info('ssh')[:user] }
+    let(:password) { conn_info('ssh')[:password] }
 
-    context "when running against puppet 5 or puppet 6" do
+    # Run tests that require the puppet-agent package to be installed on the target.
+    # Each test is run against all agent targets unless otherwise noted.
+    context 'with puppet installed' do
+      # Install agents on the agent targets.
       before(:all) do
-        # install puppet5
-        install('puppet_5', collection: 'puppet5', inventory: agent_version_inventory)
+        TEST_VERSIONS.each do |version, collection|
+          install("puppet_#{version}_node", collection: collection, inventory: docker_inventory(root: true))
+        end
+      end
 
-        result = run_task('puppet_agent::version', 'puppet_5', {}, inventory: agent_version_inventory)
-        expect(result.count).to eq(1)
-        expect(result[0]).to include('status' => 'success')
-        expect(result[0]['value']['version']).to match(/^5/)
+      # Uninstall agents after running all tests.
+      after(:all) do
+        TEST_VERSIONS.each do |version,|
+          uninstall("puppet_#{version}_node", inventory: docker_inventory(root: true))
+        end
+      end
```

Comment:
Dunno if we _really_ need to do this, but left it in since it doesn't seem to hurt. 🤷

Reply:
Hm, I don't think so - mostly because 1. It'd be annoying if there was an unrelated error uninstalling the agents that caused them to fail, 2. These containers are named such and explicitly intended to be containers with the agent installed, and for local development it'd be annoying if a test expected the agent to be installed and it wasn't.